src/ntt/mod.rs

```diff
@@ -0,0 +1,345 @@
+
+        let ark_scalars_batch = scalars_batch
+            .clone()
+            .iter()
+            .map(|v| Fr::new(v.to_ark()))
+            .collect::<Vec<Fr>>();
+        let mut ark_ntt_result = ark_scalars_batch.clone();
+
+        ark_domain.fft_in_place(&mut ark_ntt_result);
+
+        assert_ne!(ark_ntt_result, ark_scalars_batch);
+
+        // do ntt
+        // ntt_wip(&mut ntt_result, false, false, Ordering::kNN, false, batches);
+        domain.ntt(&mut ntt_result); //single ntt
+        let ntt_result_as_ark = ntt_result
+            .iter()
+            .map(|p| Fr::new(p.to_ark()))
+            .collect::<Vec<Fr>>();
+
+        assert_ne!(ntt_result, scalars_batch);
+        assert_eq!(ark_ntt_result, ntt_result_as_ark);
+
+        let mut ark_intt_result = ark_ntt_result;
+
+        ark_domain.ifft_in_place(&mut ark_intt_result);
+        assert_eq!(ark_intt_result, ark_scalars_batch);
+
+        // check that ntt output is different from input
+        assert_ne!(ntt_result, scalars_batch);
+
+        // do intt
+        let mut intt_result = ntt_result;
+
+        ntt_wip(&mut intt_result, true, false, Ordering::kNN, false, batches);
+
+        assert!(ark_intt_result == ark_scalars_batch);
+        assert!(intt_result == scalars_batch);
+
+        let mut ntt_intt_result = intt_result;
+        ntt_wip(&mut ntt_intt_result, false, false, Ordering::kNR, false, batches);
+        assert!(ntt_intt_result != scalars_batch);
+        ntt_wip(&mut ntt_intt_result, true, false, Ordering::kRN, false, batches);
+        assert!(ntt_intt_result == scalars_batch);
+
+        let mut ntt_intt_result = list_to_reverse_bit_order(&ntt_intt_result);
+        ntt_wip(&mut ntt_intt_result, false, false, Ordering::kRR, false, batches);
+        assert!(ntt_intt_result != scalars_batch);
+        ntt_wip(&mut ntt_intt_result, true, false, Ordering::kRN, false, batches);
+        assert!(ntt_intt_result == scalars_batch);
+
+        ////
+        let size = ntt_intt_result.len() / batches;
+
+        let mut config = get_ntt_config_with_input(&mut ntt_intt_result, size, batches);
+
+        ntt_internal(&mut config);
+
+        //host
+        let mut ntt_result = scalars_batch.clone();
+        ntt_wip(&mut ntt_result, false, false, Ordering::kNR, false, batches);
+
+        let mut buff1 = DeviceBuffer::from_slice(&scalars_batch[..]).unwrap();
+        let dev_ptr1 = buff1
+            .as_device_ptr()
+            .as_raw_mut();
+
+        let buff_len = buff1.len();
+
+        std::mem::forget(buff1);
+
+        let buff_from_dev_ptr = unsafe { DeviceBuffer::from_raw_parts(DevicePointer::wrap(dev_ptr1), buff_len) };
+        let mut from_device = vec![ScalarField::zero(); scalars_batch.len()];
+        buff_from_dev_ptr
+            .copy_to(&mut from_device)
+            .unwrap();
+
+        assert_eq!(from_device, scalars_batch);
+
+        // host - device - device - host
```

Comment:
mixing of host/device inputs outputs example using internal method - the ensuring safety could start at Rust side and migrate to CUDA if/where needed